image/image.go

```diff
@@ -218,6 +219,30 @@ func NewHistory(author, comment, createdBy string, isEmptyLayer bool) History {
 	}
 }
 
+// Equal compares two history structs for equality
+func (h History) Equal(i History) bool {
+	h.normalize()
+	i.normalize()
+
+	return reflect.DeepEqual(h, i)
```

Comment:
Should be cleaner to reset `Created` property and run `h.Created.Equal(i.Created)` for it specifically.

Reply:
I considered doing that, but it would mean that 14:00:00+01:00 would equal 16:00:00+03:00, for example.

Since the difference between those times is preserved across Unmarshal/Marshal (unlike "+00:00" vs "Z") I went for this approach.

Reply:
(Thinking out loud); At a quick glance, I'd say that'd be correct 🤔; if we'd treat them as "different", that would mean that setting a different timezone on a server would be considered a different artefact even if the time was actually the same.

Reply:
For a valid parent the histories being compared were copied directly from parent to child. The only difference is that the child's copies took an extra round trip through UnmarshalJSON/MarshalJSON. In that case I don't think changing the server TZ can change the result of the comparison.

That said, I'll take another look at Time.Unmarshal to make sure it isn't doing anything unexpected with the current TZ.

Reply:
@thaJeztah, here's what I found: 

- The system timezone is read once per execution and cached; the new zone isn't used until the daemon restarts
- There's a quirk in the [unmarshaling of non-local times](https://play.golang.org/p/MnqaDk-7VnL) (doesn't affect checkValidParent as it uses reflect.DeepEqual)

These points don't have any negative impact on this approach or using time.Equal instead.

To clarify, I chose this approach as it only modifies behaviour in the specific case reported in #41829. Using time.Equal would fix that issue too, but it may have knock-on effects such as failing to detect a non-valid parent.

Reply:
Thanks!

> such as failing to detect a non-valid parent.

Let me defer that one to @tonistiigi as he's more familiar with this.

We should consider consistently using UTC for these. I did a very quick check on what variants are used in the code base; looks like many paths already use `.UTC()` or `.Unix()` (which implicitly uses UTC) when reading.

I found one place that didn't use UTC; opened https://github.com/moby/moby/pull/41905 for that.

Reply:
@thaJeztah I don't know if this is important, but buildkit uses local times (time.Now()) in the history when making new images. The legacy build system uses time.Now().UTC().

Reply:
Interesting. Wondering if there was a reason for that, or if that was overlooked 🤔. For "internal" use, it probably wouldn't make a difference if `.Equal()` is used, mostly thinking about what will end up in JSON (manifests, local storage), and if _not_ using `.UTC()` would make things less reproducible/consistent. That said; fully "reproducible" is still hard, and may require things like `SOURCE_DATE_EPOCH`